### ISSUE TITLE
the progress example provided in the documentation doesn't appear to work at all for .Net framework

### ISSUE THREAD

**Issue by cdg191** [robinrodricks/FluentFTP#463]
I used your example for a .Net core 3.0 console app, and it works perfectly:

```
Progress<FtpProgress> progress = new Progress<FtpProgress>(x => {

	// When progress in unknown, -1 will be sent
	if (x.Progress < 0){
		progressBar.IsIndeterminate = true;
	}else{
		progressBar.IsIndeterminate = false;
		progressBar.Value = x;
	}
});
```

The same code, however, in a .Net 4.7.2 winform app, never fires.  The documentation says For .NET 2.0 users, pass an implementation of the IProgress class, which I ended up doing, and it works.  

This is with VS 2019.

I stepped through the latest FluentFTP code with my 4.7.2 app, and I couldn't see why the callback object never fired.  Perhaps the above example only works with .Net core, and not .Net framework?

Thanks!

**Comment by robinrodricks** [comment]
Which method are you using this with? Can you show the line of code?

**Comment by cdg191** [comment]
Oops, sorry!  It's DownloadFile.  I tried DownloadFileAsync with the same result.

```
c.DownloadFile(localFile, remoteFile, FtpLocalExists.Overwrite, FtpVerify.OnlyChecksum, progress);
```

**Comment by robinrodricks** [comment]
Maybe your file is so small it is transferred in a single chunk.

**Comment by cdg191** [comment]
No, it's not that.  I'm transferring around 42GB of data over multiple files.  Using the console version (core), the progress code fires.  Using the winform version with the same set of data, the code never fires.

**Comment by robinrodricks** [comment]
Maybe you have an older version of nuget installed with VS and so its picking up the `net20` build instead? This is a strange issue..

**Comment by robinrodricks** [comment]
Try this and see if it works?

  - New: Progress reporting for synchronous methods `Upload`, `Download`, `UploadFile` and `DownloadFile` are now sent via delegates
  - Fix: Correctly send progress for synchronous methods and retain `IProgress` for async methods

[Release.zip](https://github.com/robinrodricks/FluentFTP/files/3715760/Release.zip)

**Comment by robinrodricks** [comment]
Pushed in https://www.nuget.org/packages/FluentFTP/28.0.0

**Comment by cdg191** [comment]
This appears to work correctly now in both .Net 4.8 and .Net core.  I had to change my code in core to use:

```
Progress<FtpProgress> progress = new Progress<FtpProgress>(p =>
```
to
```
Action<FtpProgress> progress = new Action<FtpProgress>(p =>
```

Which I believe is expected?

Thanks for the quick turnaround!

**Comment by herledant** [comment]
Hello robinrodricks,

First of all many thanks to you and all the contributors for this great ftp library.
I'm facing to a problem with last version of fluentFTP (32.3.3) and the FtpProgress class.
I've try many things and it seems to not work as attending.

I try to manage a toolstripProgressbar with an ftpFluent.uploadfiles.
The progress goes end but too quick and **before the end of the uploading files**...

Here is my code...

```
Dim progress As Action(Of FtpProgress) = New Action(Of FtpProgress)(Function(x)
                                                 If x.Progress < 0 Then
                                                        'ProgressBar.IsIndeterminate = True
                                                 Else
                                                         SetControlPropertyValue1(statusprogress, "value", CInt(x.Progress))
                                                 End If
                                                 End Function)

client.UploadFiles(list, "/download/", FtpRemoteExists.Overwrite, False, FtpVerify.None, FtpError.None, progress)
client.Disconnect()
End Using
```

 ```
Delegate Sub SetControlValueCallback1(ByVal oControl As ToolStripProgressBar, ByVal propName As String, ByVal propValue As Object)

    Private Sub SetControlPropertyValue1(ByVal oControl As ToolStripProgressBar, ByVal propName As String, ByVal propValue As Object)
        If oControl.GetCurrentParent().InvokeRequired Then
            Dim d As SetControlValueCallback1 = New SetControlValueCallback1(AddressOf SetControlPropertyValue1)
            oControl.GetCurrentParent().Invoke(d, New Object() {oControl, propName, propValue})
        Else
            Dim t As Type = oControl.[GetType]()
            Dim props As PropertyInfo() = t.GetProperties()
            For Each p As PropertyInfo In props
                If p.Name.ToUpper() = propName.ToUpper() Then
                    p.SetValue(oControl, propValue, Nothing)
                End If
            Next
        End If
    End Sub
```
If you have an idea for what i'm doing wrong.

Regards

**Comment by robinrodricks** [comment]
Use async methods for better progress reporting. Its always wired behaviour with sync for some reason.

https://github.com/robinrodricks/FluentFTP/blob/master/FluentFTP.VBExamples/DownloadFileWithProgress.vb

https://github.com/robinrodricks/FluentFTP/blob/master/FluentFTP.VBExamples/UploadFileWithProgress.vb

**Comment by herledant** [comment]
Hello,

i forget to tell you that i hav already test async method.
Anyway, i change my code again with:

```
Await client.ConnectAsync(token)

Dim myProgress As Progress(Of FtpProgress) = New Progress(Of FtpProgress)(
                    Sub(p)
                        If p.Progress = 1 Then
                            ' all done!
                        Else
                            ' percent done = (p.Progress * 100)
                            SetControlPropertyValue1(statusprogress, "value", CInt(p.Progress))
                        End If
                    End Sub)

            Await client.UploadFilesAsync(list, "/download/", FtpRemoteExists.Overwrite, False, FtpVerify.None, FtpError.None, token, myProgress)
```

The problem is always there.
Async or not, this is not changing the behavior.

the progressbar goes to the end very fast but the transfert is not over.
Have you ever meet this kind of behavior?

**Comment by herledant** [comment]
Hello robinrodricks,

I continued my tests and created a new program just to test fluentFtp.
I am disappointed because there is nothing else than the FTP transfer and the behavior is the same as in my other application.
I don't uderstand what i'm going wrong.
I have tested it wit and without delegate for update the progressBar.value but it's the same...

Here my new code:

```
Imports FluentFTP
Imports System.IO
Imports System.Net
Imports System.Threading
Imports System.Threading.Tasks

Public Class Form1
    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        With OpenFileDialog1
            .Filter = "Fichier xml (*.xml)|*.xml"
            .RestoreDirectory = True

        End With

        If OpenFileDialog1.ShowDialog() = DialogResult.OK Then
            _file.Text = OpenFileDialog1.FileName
        End If
    End Sub


    Public Async Function sendFIle() As Task
        Dim token = New CancellationToken()

        Using ftp = New FtpClient()
            Dim log As FtpTrace
            'Dim listener As New TextWriterTraceListener
            log.AddListener(New TextWriterTraceListener("fluentFtp.log"))
            log.LogUserName = True
            log.LogPassword = False
            log.LogIP = True


            ftp.Host = _ip.Text
            ftp.Port = CInt(_port.Text)
            ftp.Credentials = New NetworkCredential(_userPWD.Text, _userPWD.Text)
            Await ftp.ConnectAsync(token)



            Dim nFile As String = Path.GetFileName(_file.Text)
            Dim npath As String = Path.GetDirectoryName(_file.Text)
            Dim di As New DirectoryInfo(npath)
            Dim fiarr As FileInfo() = di.GetFiles()

            Dim progress As Progress(Of FtpProgress) = New Progress(Of FtpProgress)(
                    Sub(p)
                        If p.Progress = 1 Then
                            ' all done!
                        Else
                            ' percent done = (p.Progress * 100)
                            'ProgressBar1.Value = p.Progress
                            SetControlPropertyValue1(ProgressBar1, p.Progress)
                        End If
                    End Sub)
            Dim sourceList As New List(Of String)

            'Try
            Dim i As Integer = 0
            For Each f In fiarr
                i = i + 1
                'MsgBox(f.FullName)
                sourceList.Add(f.FullName)

            Next

            Await ftp.UploadFilesAsync(sourceList, "/download/", FtpRemoteExists.Overwrite, False, FtpVerify.None, FtpError.None, token, progress)
            'MsgBox("transfert ok: " & i & " file(s)")
            'Catch ex As Exception
            ' MsgBox("Error: " & ex.Message)
            'End Try


            'Await ftp.DisconnectAsync()
        End Using


    End Function

    Delegate Sub SetControlValueCallback1(ByVal oControl As ProgressBar, ByVal text As String)



    Private Sub SetControlPropertyValue1(ByVal ctl As ProgressBar, ByVal text As String)
        If ctl.InvokeRequired Then
            ctl.Invoke(New SetControlValueCallback1(AddressOf SetControlPropertyValue1),
                       ctl,
                       text)
        Else
            ctl.Value = text
        End If
    End Sub

    Private Async Function Button2_ClickAsync(sender As Object, e As EventArgs) As Task Handles Button2.Click
        Await sendFIle()

    End Function
End Class
```

**Comment by herledant** [comment]
Hello robinrodricks,

Sorry to insist but, i really think there is a problem with the ftpProgress.progress.

I check the value and it always equal to 100. That's why, the progressBar goes to end at the first uploading file.
See my screenshot when for each file i show the value of FtpProgress.progress value:

![2020-04-17_15h16_55](https://user-images.githubusercontent.com/9166449/79573474-00dfd780-80bf-11ea-9560-78c863fd0bb4.png)

Thank's for your help.

**Comment by tomdewilde** [comment]
@herledant  when uploading multiple files you should look at FileCount & FileIndex of FtpProgress. The Progress property is 'per file' and thus might be always 100 when uploading 'small' files

**Comment by Wiilf** [comment]
Sorry to drop in so late, I just got this library last night and I had nothing but issues with the progressbar, and my situation being a custom drawn progressbar with update/add methods for a listview. 

I tried different formulas to calculate the percentage values, a lot of them were off.. some close, but overall nothing but overflows. 

As it turns out, pulling some calc code from a separate part of my project, and added to this instance, it calculates 0-100 perfectly. I've tested a dozen times on all kinds of file sizes. I really appreciate the note in the _TransferredBytes_ meta description! I dont think I would of pushed on too much further if I didnt notice that :D

`double x = (int)(file.TransferredBytes * 100 / filesize);
`
`
int v = (int)Math.Round(x) > 100 ? 100 : (int)Math.Round(x);
`

In case this helps anyone. I use a static class for Math rounding that is high performing on x64 processors, look around on SO for "FastMath" as to reduce this potential overhead. 

Great library btw! I've implemented it fairly well into our project and I cant thank you enough.
